### DISCUSSION TITLE
Multi-part File Upload for large files

### DISCUSSION THREAD

**Discussion by elockman** [babelouest/ulfius#203] [General]
I would like to upload large files with chunking.  The sheep_counter provides a good start.  It uploads files with maximum chunk sizes of 16kb.  I would like to know how to determine when the last chunk is delivered, in order to invoke code to piece all of the chunks together, in order to reassemble the large file.

In the sheep_counter example, **ulfius_set_upload_file_callback_function**() assigns **file_upload_callback**() to intercept the chunked data nicely. 

Is there a function that provides total file size, or number of expected chunks that I'm missing?  Is there anything that would notify the application code that the last chunk is being sent?

**Comment by elockman** [comment]
My current implementation is to do incremental reconstructions after each chunk.  This avoids needing to know when the last chunk is received.  Each chunk is concatenated as a binary file inside the **file_upload_callback**().    By treating each chunk as binary files, I am able to reconstruct all file types.

**Comment by babelouest** [comment]
Unfortunately, the `file_upload_callback` can't know the full size of the uploaded file before the last chunk.

Also because an HTTP request doesn't require a `content-length` header, so if the client sets this header, you may guess the total file size, but otherwise I don't see how...

**Comment by elockman** [comment]
Yes, that makes sense.  The incremental reconstruction approach seems to be working well.
Thank you for your response.
